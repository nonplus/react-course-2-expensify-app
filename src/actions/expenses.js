import uuid from "uuid";
import database from "../firebase/firebase";
import _ from "lodash";

export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const addExpenseAsync = (expenseData = {}) => {
  return async dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    const ref = await database.ref("expenses").push(expense);
    await dispatch(addExpense({ id: ref.key, ...expense }));
  };
};

export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const setExpenseAsync = () => {
  return async dispatch => {
    const snapshot = await database.ref("expenses").once("value");
    const expenses = _.map(snapshot.val(), (value, id) => ({ id, ...value }));
    dispatch(setExpenses(expenses));
  };
};

export const editExpense = (id, updates) => {
  const { description, note, amount, createdAt } = updates;
  return {
    type: "EDIT_EXPENSE",
    id,
    updates: {
      description,
      note,
      amount,
      createdAt
    }
  };
};

export const editExpenseAsync = (id, updates) => {
  const { description, note, amount, createdAt } = updates;
  return async dispatch => {
    await database
      .ref(`expenses/${id}`)
      .set({ description, note, amount, createdAt });
    dispatch(editExpense(id, updates));
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const removeExpenseAsync = ({ id } = {}) => {
  return async dispatch => {
    await database.ref(`expenses/${id}`).remove();
    dispatch(removeExpense({ id }));
  };
};
