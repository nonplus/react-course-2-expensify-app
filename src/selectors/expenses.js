export const selectExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  const startDateMatches = startDate
    ? expense => startDate.isSameOrBefore(expense.createdAt, "day")
    : () => true;
  const endDateMatches = endDate
    ? expense => endDate.isSameOrAfter(expense.createdAt, "day")
    : () => true;
  text = text.toLowerCase();
  const textMatches = !text
    ? () => true
    : expense => expense.description.toLowerCase().includes(text);

  const sortCompare =
    sortBy === "date"
      ? (a, b) => b.createdAt - a.createdAt
      : sortBy === "amount" ? (a, b) => b.amount - a.amount : () => 0;

  return expenses
    .filter(
      expense =>
        startDateMatches(expense) &&
        endDateMatches(expense) &&
        textMatches(expense)
    )
    .sort(sortCompare);
};
