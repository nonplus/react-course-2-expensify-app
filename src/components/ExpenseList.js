import React from "react";
import { connect } from "react-redux";
import { selectExpenses } from "../selectors/expenses";
import ExpenseListItem from "./ExpenseListItem";

export const ExpenseList = ({ expenses }) => (
  <div className="content-container">
    <div className="list-header">
      <div className="mobile-only">Expenses</div>
      <div className="desktop-only">Expense</div>
      <div className="desktop-only">Amount</div>
    </div>
    <div className="list-body">
      {expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No expenses.</span>
        </div>
      ) : (
        expenses.map(expense => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )}
    </div>
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
