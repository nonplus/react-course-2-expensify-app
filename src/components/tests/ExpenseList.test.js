import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../ExpenseList";
import expenses from "../../tests/fixtures/expenses";

describe("<ExpenseList />", () => {
  it("should remder list with expenses", () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should show default message with empty expenses", () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
  });
});
