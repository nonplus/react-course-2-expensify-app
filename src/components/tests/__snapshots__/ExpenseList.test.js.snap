// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList should remder list with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="id_1"
    key="id_1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="id_2"
    key="id_2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="id_3"
    key="id_3"
    note=""
  />
</div>
`;

exports[`ExpenseList should show default message with empty expenses 1`] = `
<div>
  <p>
    No expenses.
  </p>
</div>
`;
